nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/QuerySalesforceObject.java

```diff
@@ -0,0 +1,492 @@
+                            .valid(false)
+                            .explanation("it requires " + AGE_FIELD.getDisplayName() + " also to be set.")
+                            .build()
+            );
+        }
+        return results;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        String sObject = context.getProperty(SOBJECT_NAME).getValue();
+        String fields = context.getProperty(FIELD_NAMES).getValue();
+        String customWhereClause = context.getProperty(CUSTOM_WHERE_CONDITION).getValue();
+        RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean();
+
+        String ageField = context.getProperty(AGE_FIELD).getValue();
+        String initialAgeFilter = context.getProperty(INITIAL_AGE_FILTER).getValue();
+        Long ageDelayMs = context.getProperty(AGE_DELAY).asTimePeriod(TimeUnit.MILLISECONDS);
+
+        String ageFilterLower;
+        StateMap state;
+        try {
+            state = context.getStateManager().getState(Scope.CLUSTER);
+            ageFilterLower = state.get(LAST_AGE_FILTER);
+        } catch (IOException e) {
+            throw new ProcessException("Last Age Filter state retrieval failed", e);
+        }
+
+        String ageFilterUpper;
+        if (ageField == null) {
+            ageFilterUpper = null;
+        } else {
+            Instant ageFilterUpperTime;
+            if (ageDelayMs == null) {
+                ageFilterUpperTime = Instant.now();
+            } else {
+                ageFilterUpperTime = Instant.now().minus(ageDelayMs, ChronoUnit.MILLIS);
+            }
+            ageFilterUpper = DateTimeFormatter.ISO_OFFSET_DATE_TIME
+                    .withZone(ZoneId.systemDefault())
+                    .format(ageFilterUpperTime);
+        }
+
+        String describeSObjectResult = salesforceRestService.describeSObject(sObject);
+        ConvertedSalesforceSchema convertedSalesforceSchema = convertSchema(describeSObjectResult, fields);
+
+        String querySObject = buildQuery(
+                sObject,
+                fields,
+                customWhereClause,
+                ageField,
+                initialAgeFilter,
+                ageFilterLower,
+                ageFilterUpper
+        );
+        String querySObjectResult = salesforceRestService.query(querySObject);
+
+        FlowFile flowFile = session.create();
+
+        Map<String, String> originalAttributes = flowFile.getAttributes();
+        Map<String, String> attributes = new HashMap<>();
+
+        AtomicInteger recordCountHolder = new AtomicInteger();
+
+        flowFile = session.write(flowFile, out -> {
+            InputStream querySObjectResultInputStream = new ByteArrayInputStream(querySObjectResult.getBytes());
```

Comment:
This conversion from String to byte array implicitly uses the platform character set, which could be problematic if the response string. In addition, this conversion from String to byte array to JSON-based records could result in notable memory consumption for larger result sets.  It looks like the `SalesforceRestService` should be refactored to avoid reading the entire response into memory.  Instead, the response stream could be read directly using the JSON Reader.